File: 02linear_list/01array/214_Search_in_Rotated_Sorted_Array_II/main.py
#!/usr/bin/env python
# coding=utf-8


class Solution:
        # @param A, a list of integers
        # @param target, an integer to be searched
        # @return an integer
        def search(self, A, target):
            lo = 0
            hi = len(A)-1
            
            while lo <= hi:
                mid = (lo + hi)/2
                if target == A[mid]:
                    return True

                if A[lo] < A[mid]:
                    if target < A[mid] and target >= A[lo]:
                        hi = mid
                    else:
                        lo = mid+1
                elif A[lo] > A[mid]:
                    if target > A[mid] and target <= A[hi]:
                        lo = mid+1
                    else:
                        hi = mid
                else:
                    lo = lo+1

            return False


if __name__ == "__main__":
    s = Solution()
    A = [1,3]
    B = [7,8,9,1,2,3,4]
    C = [1]

    print s.search(A,3)
    print s.search(B,8)
    print s.search(C,1)

File: 02linear_list/01array/214_Search_in_Rotated_Sorted_Array_II/main.cpp
/*************************************************************************
	> File Name: main.cpp
	> Author:zhaoyihong 
	> Mail:zhaoyihong_at_126_dot_com 
	> Created Time: 2015年03月16日 星期一 17时28分16秒
 ************************************************************************/

#include  <iostream>
using namespace std;


class Solution {
public:
    bool search(int A[], int n, int target) {
        
        int lo = 0 ,hi = n-1;
        
        while(lo <= hi)
        {
            int mid = (lo + hi) / 2;
            
            if(target == A[mid])
            {
                return true;
            }
            
            if(A[lo] < A[mid])
            {
                //左边有序的情况

                //判断目标是否在左边的区间中
                if(target < A[mid] && target >= A[lo])
                {
                    hi = mid;
                }
                else
                {
                    lo = mid + 1;
                }

            }
            else if(A[lo] > A[mid])
            {
                //左边无序,说明右边有序
                
                //判断target是否在右边的区间中
                if(target > A[mid]  && target <= A[hi])
                {
                    lo = mid+1;
                }
                else
                {
                    hi = mid;
                }

            }
            else
            {
                lo++;
            }
        }

        return false;

    }
};


int main()
{
    int A[] = {
        2,2,2,3,3,1,1,2
    };

    cout << boolalpha <<  Solution().search(A,sizeof(A)/sizeof(*A),2) << endl;
    cout << boolalpha <<  Solution().search(A,sizeof(A)/sizeof(*A),3) << endl;
    cout << boolalpha <<  Solution().search(A,sizeof(A)/sizeof(*A),1) << endl;
}
